compiler/lib/src/main/scala/codegen/CppWriter/ComponentCppWriter/ComponentEvents.scala

```diff
@@ -271,6 +284,26 @@ case class ComponentEvents (
               |_id = this->getIdBase() + ${eventIdConstantName(event.getName)};
               |"""
         ),
+        // Time based throttle timeout needs above time
+        event.throttle match {
+          case Some(Event.Throttle(_, Some(Event.TimeInterval(seconds, useconds)))) => lines(
+            s"""|// Check throttle value & throttle timeout
+                |if (this->${eventThrottleCounterName(event.getName)} == 0 ||
```

Comment:
This code is not concurrency safe. The original code was doing a fetch and increment. We need to figure out how to add this logic to the fetch and increment.

Reply:
> I think the existing logic for checking the throttle count also needs to be corrected.

The existing logic is
```
1  if (throttleCount < maxThrottleCount) {
2    atomically increment throttle count
3    emit event
4  }
```
That's not correct because any number of threads could read the same throttle count value in line 1. Instead I think it should be
```
1  if (throttleCount < maxThrottleCount) {
2    atomically increment throttleCount and set prevThrottleCount to the value before the increment
3    if (prevThrottleCount < maxThrottleCount) {
4      emit event
5    }
6  }
```
That way any number of threads can read the throttle count in line 1, but only one of them will win the race in line 2 and proceed to line 4.